### PR TITLE
WIP: make alglist return the list of algorithms

### PR DESCRIPTION
**Pull request by mbernasocchi**
This would allow checking for the presence of algorithm
```
l = processing.alglist()
if 'myNeededAlg' in l:
  do_things()
```

### CODE REVIEW

**Comment by m-kuhn** [comment]
I think there should be two different methods:

 * one that returns a list of algorithm names
 * one that prints a '\n' delimited string with all algs (I don't know for what, but that's what the method currently does, where is it used now?)

**Comment by alexbruy** [comment]
@m-kuhn this method (one that prints '\n' delimited list of algorithms) is a helper method and mainly used for looking for available algs. AFAIK is not used in the Processing code.

**Comment by m-kuhn** [comment]
@alexbruy is it expected to be invoked manually (from the console)?

What about renaming it to `dumpAlglist()` or `printAlglist()` and have `alglist()` return a list?

**Comment by alexbruy** [comment]
@m-kuhn correct, it is usually invoked manually by user. Renaming sounds good for me, this will help to avoid possible confusion.

@volaya any objections?

**Comment by volaya** [comment]
Sounds good to me. We can have alglist to return a list object, and then printalglist to print in the console. A user that is used to the previous method wont get an error, but the list printed instead (not so nice, but it's something...and he can then find about the new method in the docs)

Before merging this (with those changes proposed), let's make sure that we have a corresponding PR in the docs repo. Otherwise, we will most likely forget to document it...

**Comment by m-kuhn** [comment]
@mbernasocchi feel like doing a little python today?

**Comment by mbernasocchi** [comment]
ahhaha sorry I'm in JS mood :)

**Comment by m-kuhn** [comment]
@mbernasocchi just delete line 52 and we have a shorter pull request queue? ;)

**Comment by mbernasocchi** [comment]
doing a better job :)

**Comment by m-kuhn** [comment]
Thanks.

In QGIS there is (almost) no `get`-prefix. Can we keep this semantic and instead of tagging the getters with `get` tag the printers with `print`?

**Comment by mbernasocchi** [comment]
@m-kuhn have a look if you like it and if I broke anything

**Comment by mbernasocchi** [comment]
was a quick fix so you'r happy, need to go back to JS

File: python/plugins/processing/tools/general.py [hunk 1]
```diff
@@ -42,17 +42,36 @@
...
+def get_alglist(text=None):
     s = ''
     for provider in list(algList.algs.values()):
         sortedlist = sorted(list(provider.values()), key=lambda alg: alg.name)
         for alg in sortedlist:
             if text is None or text.lower() in alg.name.lower():
                 s += alg.name.ljust(50, '-') + '--->' + alg.commandLineName() \
                     + '\n'
-    print(s)
+    return s
```
Comment (m-kuhn) [inline python/plugins/processing/tools/general.py]: Does this return a string now?
Wouldn't it be more appropriate to return a `list`?

**Comment by mbernasocchi** [comment]
@m-kuhn renaming the original maethods to print is not an option for me right now due to time constrains, there are 40 occurences of alglist in processing and I'd need to check what these calls are. plus I'd have to document the change. I can for sure have a look at this eventually, just definetly not today

File: python/plugins/processing/tools/general.py [hunk 2]
```diff
@@ -42,17 +42,36 @@
...
+    l = []
     for provider in list(algList.algs.values()):
         sortedlist = sorted(list(provider.values()), key=lambda alg: alg.name)
         for alg in sortedlist:
             if text is None or text.lower() in alg.name.lower():
-                s += alg.name.ljust(50, '-') + '--->' + alg.commandLineName() \
-                    + '\n'
-    print(s)
+                t = alg.name.ljust(50, '-') + '--->' + alg.commandLineName()
+                l.append(t)
```
Comment (m-kuhn) [inline python/plugins/processing/tools/general.py]: Could this return a dict?

    { alg.name: alg.commandLineName() for alg in sortedList }

And the formatting would happen in the print method?

The current separation of data and representation confuses me a bit.

**Comment by m-kuhn** [comment]
> there are 40 occurences of alglist in processing

fair enough, I thought there are none based on @alexbruy comments above

**Comment by m-kuhn** [comment]
I'll leave a few comments on what I think would be good to check as well when you get to it :)

File: python/plugins/processing/tools/general.py [hunk 3]
```diff
@@ -42,17 +42,36 @@
...
+    
+
+def alghelp(name):
+    s = get_alghelp(name)
+    if s:
+        print(s)
+    else:
+        print('Algorithm not found')
+    
+def get_alglist(text=None):
```
Comment (m-kuhn) [inline python/plugins/processing/tools/general.py]: would be great to change `text` to something like `algname`

**Comment by mbernasocchi** [comment]
> fair enough, I thought there are none based on @alexbruy comments above
not sure, maybe there is another method called alglist.

File: python/plugins/processing/tools/general.py (hunk 2 above)
Comment (mbernasocchi) [inline python/plugins/processing/tools/general.py]: which python version do we need to support? dict are not ordered, could use OrderedDict from 2.7

**Comment by alexbruy** [comment]
@mbernasocchi if you don't mind, I will try to find time to work on this next week

**Comment by mbernasocchi** [comment]
> if you don't mind
not at all :)

**Comment by nyalldawson** [comment]
Superseded by #4293

**Pull request closed without merging**